server/welcomebot.go

```diff
@@ -228,12 +230,55 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
 }
 
 func (p *Plugin) joinChannel(action *Action, channelName string) {
```

Comment:
I'd prefer to have this function split up a bit.  Something like the following: 
```go 
	// If it begins with @ create a DM channel
	if strings.HasPrefix(channelName, "@") {
       doSomethingForDMs()
       return
    } 

    // Otherwise treat it like a normal channel
    if channel, err := p.API.GetChannelByName(action.Context.TeamID, channelName, false); err == nil {
		if _, err := p.API.AddChannelMember(channel.Id, action.Context.UserID); err != nil {
			p.API.LogError("Couldn't add user to the channel, continuing to next channel", "channel_name", channelName, "user_id", action.Context.UserID, channel.Id)
			return
		}
	} else {
		p.API.LogError("failed to get channel, continuing to the next channel", "channel_name", channelName, "user_id", action.Context.UserID)
		return
	}
```

Reply:
This should make the code a bit cleaner and the PR will only show changes that are specific to the new functionality because the diff won't be changed for the existing, unchanged code because of formatting.

Reply:
On top of this, I recommend having the function `doSomethingForDMs` return an error, and have the caller log the error. To include extra information when returning an error, you can use `errors.Wrap` or `errors.Wrapf`:

```go
err := doSomethingForDMs(...)
if err != nil {
    p.API.LogError("failed to handle DM channel, continuing to next channel. " + err.Error())
}

// Then in doSomethingForDMs()
func doSomethingForDMs(username string, ...) error {
    dmUser, userErr := p.API.GetUserByUsername(username)
    if userErr != nil {
        return errors.Wrapf(userErr, "couldn't find DM channel for username %s", username)
    }
    // ...
}
```

Doing this decouples the error creation and logging functionality